openstudiocore/src/model/FoundationKiva.cpp

```diff
@@ -309,6 +340,94 @@ namespace detail {
     return getObject<ModelObject>().getModelObjectSources<Surface>(Surface::iddObjectType());
   }
 
+  unsigned int FoundationKiva_Impl::numberofCustomBlocks() const {
+    return numExtensibleGroups();
+  }
+
+  bool FoundationKiva_Impl::setNumberofCustomBlocks(unsigned int numberofCustomBlocks) {
+    bool result = setInt(OS_Foundation_KivaFields::NumberofCustomBlocks, numberofCustomBlocks);
+    return result;
+  }
+
+  void FoundationKiva_Impl::resetNumberofCustomBlocks() {
+    bool result = setInt(OS_Foundation_KivaFields::NumberofCustomBlocks, 0);
+    OS_ASSERT(result);
+  }
+  
+  bool FoundationKiva_Impl::addCustomBlock(const CustomBlock& customBlock) {
+    bool result;
+   
+    unsigned int num = numberofCustomBlocks();
+    // Max number of custom blocks is 10
+    if (num >= 10) {
+      LOG(Warn, briefDescription() << " already has 10 custom blocks which is the limit");
+      result = false;
+    } else {
+      // Push an extensible group
+      WorkspaceExtensibleGroup eg = getObject<ModelObject>().pushExtensibleGroup().cast<WorkspaceExtensibleGroup>();
+      bool material = eg.setPointer(OS_Foundation_KivaExtensibleFields::CustomBlockMaterialName, customBlock.material().handle());
+      bool depth = eg.setDouble(OS_Foundation_KivaExtensibleFields::CustomBlockDepth, customBlock.depth());
+      bool xPosition = eg.setDouble(OS_Foundation_KivaExtensibleFields::CustomBlockXPosition, customBlock.xPosition());
+      bool zPosition = eg.setDouble(OS_Foundation_KivaExtensibleFields::CustomBlockZPosition, customBlock.zPosition());
+      if (material && depth && xPosition && zPosition) {
+        setNumberofCustomBlocks(num + 1);
+        result = true;
+      } else {
+        // Something went wrong
+        // So erase the new extensible group
+        getObject<ModelObject>().eraseExtensibleGroup(eg.groupIndex());
+        result = false;
+      }
+    }    
+    return result;
+  }
+  
+  bool FoundationKiva_Impl::addCustomBlock(const Material& material, double depth, double xPosition, double zPosition) {
+    // Make a custom block, and then call the above function
+    CustomBlock customBlock(material, depth, xPosition, zPosition);
+    return addCustomBlock(customBlock);
+  }
+  
+  bool FoundationKiva_Impl::removeCustomBlock(unsigned groupIndex) {
```

Comment:
I have tended to not include this method, it is hard to use.  I think its easier to just clear all the blocks and then set all of them with a vector.  If you keep this method it should have a lot of tests

Reply:
Decided to keep this method and have a lot of tests.